### PR TITLE
FOGL-556 Implement Backup API

### PR DESCRIPTION
**Pull request by Singhal-Vaibhav**
**Done:**

| Verb | Endpoint |
| --- | --- |
| GET, POST | `/foglamp/backup` |
| GET, DELETE | `/foglamp/backup/{backup-id}` |

Tests for the above endpoints.

**TODO's**

- [FOGL-865](https://scaledb.atlassian.net/browse/FOGL-865)
- [FOGL-861](https://scaledb.atlassian.net/browse/FOGL-861)

**Test Results**
```
pytest -s -vv test_backup_restore.py 
================================================================================= test session starts ==================================================================================
platform linux -- Python 3.5.2, pytest-3.1.1, py-1.4.34, pluggy-0.4.0 -- /usr/bin/python3
cachedir: .cache
rootdir: /home/foglamp/Development/FogLAMP/tests/unit-tests/python/foglamp_test/services/core/api, inifile:
plugins: mock-1.6.0, cov-2.5.1, asyncio-0.6.0, allure-adaptor-1.7.7
collected 15 items 

test_backup_restore.py::TestBackup::test_get_backups_positive_cases[-200-4-exp_output0] PASSED
test_backup_restore.py::TestBackup::test_get_backups_positive_cases[?limit=1-200-1-exp_output1] PASSED
test_backup_restore.py::TestBackup::test_get_backups_positive_cases[?skip=3-200-1-exp_output2] PASSED
test_backup_restore.py::TestBackup::test_get_backups_positive_cases[?limit=2&skip=1-200-2-exp_output3] PASSED
test_backup_restore.py::TestBackup::test_get_backups_positive_cases[?status=failed-200-1-exp_output4] PASSED
test_backup_restore.py::TestBackup::test_get_backups_positive_cases[?limit=2&skip=1&status=completed-200-1-exp_output5] PASSED
test_backup_restore.py::TestBackup::test_get_backups_negative[?limit=invalid-200-400-limit must be an integer] PASSED
test_backup_restore.py::TestBackup::test_get_backups_negative[?skip=invalid-200-400-skip must be an integer] PASSED
test_backup_restore.py::TestBackup::test_get_backups_negative[?status=invalid-200-400-'INVALID' not a valid status] PASSED
test_backup_restore.py::TestBackup::test_get_backup_details_positive[request_params0-200-output0] PASSED
test_backup_restore.py::TestBackup::test_get_backup_details_negative[invalid-200-400-Invalid backup id] PASSED
test_backup_restore.py::TestBackup::test_get_backup_details_negative[-1-200-404-Backup with -1 does not exist] PASSED
test_backup_restore.py::TestBackup::test_create_backup SKIPPED
test_backup_restore.py::TestBackup::test_delete_backup SKIPPED
test_backup_restore.py::TestRestore::test_restore SKIPPED
```

### CODE REVIEW

**Comment by Singhal-Vaibhav** [comment]
@sstefano69 Can you please review?

File: python/foglamp/services/core/api/backup_restore.py [hunk 1]
```diff
@@ -7,113 +7,148 @@
...
+from collections import OrderedDict
+
+from foglamp.services.core import connect
+from foglamp.plugins.storage.postgres.backup_restore.backup_postgres import Backup
+from foglamp.plugins.storage.postgres.backup_restore import exceptions
+
 # TODO: remove this and call actual class methods
 from unittest.mock import MagicMock
-Backup = MagicMock()
+Backup_Mock = MagicMock()
```
Comment (MarkRiddoch) [inline python/foglamp/services/core/api/backup_restore.py]: Clearly we can not merge with this still in place

Comment (Singhal-Vaibhav) [reply python/foglamp/services/core/api/backup_restore.py]: Ok. We can do it either way:
1. Pull FOGL-861 and FOGL-834 in current sprint and implement the mocked methods in this pull only.
2. Remove routes for restore and subsequently the restore api code with a TODO to be addressed in a future sprint. That way we'll have SOME logic for restore stuff but not accessible by api.

**Comment by sstefano69** [comment]
I also tried the code (some executions) and I saw that the 6=Restored is shown as UNKNOWN : 

'foglamp@foglamp:~/Development/FogLAMP$ psql -t -c "SELECT id,file_name,status FROM foglamp.backups;"'
' 30 | /usr/local/foglamp/data/backup/foglamp_backup_2017_12_20_09_55_47.dump                |      6'
 
'foglamp@foglamp:~/Development/FogLAMP$ curl -X GET http://localhost:8081/foglamp/backup/30
{"id": 30, "status": "UNKNOWN", "date": "2017-12-20' '08:55:48.001855+00"}foglamp@foglamp:~/Development/FogLAMP$'

File: python/foglamp/services/core/api/backup_restore.py [hunk 2]
```diff
@@ -7,113 +7,148 @@
...
+    except exceptions.DoesNotExist:
+        raise web.HTTPNotFound(reason='Backup with {} does not exist'.format(backup_id))
+    except Exception as ex:
+        raise web.HTTPException(reason=str(ex))
+
 
-        _resp = Backup.delete_backup(id=backup_id)
-        return web.json_response({'message': _resp})
+def _get_status(status_code):
+    if status_code not in range(1, 6):
```
Comment (ashish-jabble) [inline python/foglamp/services/core/api/backup_restore.py]: In your case below case fails:
```
if 6 not in range(1, 6):
    print ("YES")
```    
YES

So, range check should (1, 7 ) as per Status enum

File: tests/unit-tests/python/foglamp_test/services/core/api/test_backup_restore.py [hunk 3]
```diff
@@ -46,51 +51,45 @@
...
-async def teardown_module(module):
+def teardown_module(module):
     """
-    Delete the created files from backup directory
+    Delete the created files from backup db, directory (if created)
     """
-    # FIXME after actual implementation
-    # asyncio.get_event_loop().run_until_complete(delete_master_data())
+    asyncio.get_event_loop().run_until_complete(delete_master_data())
     pass
```
Comment (ashish-jabble) [inline tests/unit-tests/python/foglamp_test/services/core/api/test_backup_restore.py]: :) pass

File: python/foglamp/services/core/api/backup_restore.py [hunk 4]
```diff
@@ -7,113 +7,148 @@
...
-
-        # backup_json = [{"id": b[0], "date": b[1], "status": b[2]}
-        #                for b in Backup.get_backup_list(limit=limit, skip=skip, status=status)]
-        backup_json = Backup.get_backup_list(limit=limit, skip=skip, status=status)
-    except Backup.DoesNotExist:
-        raise web.HTTPNotFound(reason='No backups found for queried parameters')
-    return web.json_response({"backups": backup_json})
+        backup = Backup(connect.get_storage())
+        limit = int(request.query['limit']) if 'limit' in request.query else __DEFAULT_LIMIT
+        skip = int(request.query['skip']) if 'skip' in request.query else __DEFAULT_OFFSET
```
Comment (ashish-jabble) [inline python/foglamp/services/core/api/backup_restore.py]: What if limit and offset with bad data?

Query: `http://localhost:8081/foglamp/backup?limit=blah`
Right now it throws ValueError Exception (`"message": "invalid literal for int() with base 10: 'blah'",`), but I think we should raise like below to make more appropriate message.

```
 if 'limit' in request.query:
        if request.query.get('limit').isdigit():
            limit = int(request.query.get('limit'))
 else:
       raise web.HTTPBadRequest(reason="limit must be an integer")
```

File: python/foglamp/services/core/api/backup_restore.py (hunk 1 above)
Comment (MarkRiddoch) [reply python/foglamp/services/core/api/backup_restore.py]: We do not have the bandwidth to pull these other tickets into the sprint, Let's pull the route out and push with the mocks. I don't like the idea of having mocked things in the repository.

**Comment by Singhal-Vaibhav** [comment]
@sstefano69 Fixed, you were facing the same issue for which @ashish-jabble told the reason.
You will get {"status": "RESTORED"} in that case now.

File: tests/unit-tests/python/foglamp_test/services/core/api/test_backup_restore.py [hunk 5]
```diff
@@ -100,29 +96,86 @@ class TestBackup:
...
+            assert retval['backups'][i]['status'] == Status(exp_output[exp_length - count]["status"]).name
+            assert retval['backups'][i]['date'] is not None
+        conn.close()
+
+    @pytest.mark.parametrize("request_params, response_code, output_code, output_message", [
+        ('?limit=invalid', 200, 400, "limit must be an integer"),
+        ('?skip=invalid', 200, 400, "skip must be an integer"),
+        ('?status=invalid', 200, 400, "'INVALID' not a valid status"),
+    ])
+    async def test_get_backups_negative(self, request_params, response_code, output_code, output_message):
```
Comment (ashish-jabble) [inline tests/unit-tests/python/foglamp_test/services/core/api/test_backup_restore.py]: Better name could be test_get_backups_invalid_request or test_get_backups_bad_data

File: tests/unit-tests/python/foglamp_test/services/core/api/test_backup_restore.py [hunk 6]
```diff
@@ -100,29 +96,86 @@ class TestBackup:
...
         retval = json.loads(r)
-        assert output == retval
+        response_length = len(retval['backups'])
+        assert exp_length == response_length
+        count = 0
+        for i in range(response_length):
+            count += 1
+            assert retval['backups'][i]['id'] == exp_output[exp_length - count]['id']
+            assert retval['backups'][i]['status'] == Status(exp_output[exp_length - count]["status"]).name
+            assert retval['backups'][i]['date'] is not None
```
Comment (ashish-jabble) [inline tests/unit-tests/python/foglamp_test/services/core/api/test_backup_restore.py]: Please be consistent either `actual == expected or expected == actual`

File: tests/unit-tests/python/foglamp_test/services/core/api/test_backup_restore.py [hunk 7]
```diff
@@ -100,29 +96,86 @@ class TestBackup:
...
+        assert output['id'] == retval['id']
+        assert Status(output['status']).name == retval['status']
+        assert retval['date'] is not None
+        conn.close()
+
+    @pytest.mark.parametrize("request_params, response_code, output_code, output_message", [
+        ('invalid', 200, 400, "Invalid backup id"),
+        ('-1', 200, 404, "Backup with -1 does not exist")
+    ])
+    async def test_get_backup_details_negative(self, request_params, response_code, output_code, output_message):
```
Comment (ashish-jabble) [inline tests/unit-tests/python/foglamp_test/services/core/api/test_backup_restore.py]: Better name could be test_get_backup_details_invalid

File: python/foglamp/services/core/api/backup_restore.py [hunk 8]
```diff
@@ -7,129 +7,177 @@
...
 
 async def restore_backup(request):
     """
     Restore from a backup
-
-    :Example: curl -X PUT  http://localhost:8082/foglamp/backup/1/restore
+    :Example: curl -X PUT http://localhost:8081/foglamp/backup/1/restore
     """
+
+    raise web.HTTPNotImplemented(reason='restore backup method is not implemented yet.')
```
Comment (ashish-jabble) [inline python/foglamp/services/core/api/backup_restore.py]: Restore..

File: python/foglamp/services/core/api/backup_restore.py (hunk 1 above)
Comment (Singhal-Vaibhav) [reply python/foglamp/services/core/api/backup_restore.py]: Mocks removed. calling the restore endpoint raises exception "Restore backup method is not implemented yet"

**Comment by sstefano69** [comment]
the status handling is ok.

personally, I would leave the restore function working, so uncommented, 
but let @MarkRiddoch decide about that.

it is fine apart from the restore consideration.

**Comment by Singhal-Vaibhav** [comment]
@sstefano69 Regarding commented code in restore API, right now we will display "method not implemented" as it is wrong to restore on latest ID when ID is passed by user. We will take care of it in FOGL-861 and when FOGL-834 unblocks it.

**Pull request merged**